Implement the logic for returning to the Analytics or Key Metrics ste…

ankitrox opened the pull request:
…p when exiting the setup flow.

## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #11388

## Relevant technical choices

<!-- Please describe your changes. -->

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 5.2 and PHP 7.4.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [x] Run the code.
- [x] Ensure the acceptance criteria are satisfied.
- [x] Reassess the implementation with the IB.
- [x] Ensure no unrelated changes are included.
- [x] Ensure CI checks pass.
- [x] Check Storybook where applicable.
- [x] Ensure there is a QA Brief.
- [x] Ensure there are no unexpected significant changes to file sizes.

## Merge Reviewer Checklist

- [x] Ensure the PR has the correct target branch.
- [x] Double-check that the PR is okay to be merged.
- [x] Ensure the corresponding issue has a ZenHub release assigned.
- [x] Add a changelog message to the issue.

hussain-t commented on tests/phpunit/integration/Core/Admin/ScreenTest.php:
```diff
@@ -124,4 +125,199 @@ public function test_enqueue_assets() {
+						if ( $is_analytics_connected ) {
+							wp_safe_redirect(
+								$ctx->admin_url(
+									'user-input',
```
Same here.
```suggestion
									'key-metrics-setup',
```

hussain-t commented on includes/Core/Admin/Screens.php:
```diff
@@ -359,16 +371,62 @@ private function get_screens() {
+										$context->admin_url(
+											'user-input',
+											array(
+												'showProgress' => true,
```
Using a boolean `('showProgress' => true)` value is incorrect, which gets converted to `?showProgress=1` in the URL. However, the JavaScript code checks for the string `'true'`:
https://github.com/google/site-kit-wp/blob/3d1320f43907bd2ff01073cc89cc17929ce20172/assets/js/components/setup/ModuleSetup.js#L118-L121
This means the progress indicator won't display because `"1" === 'true'` evaluates to `false`. We should change to `'showProgress' => 'true'` (string) to match the pattern used in `Analytics_4.php`:
https://github.com/google/site-kit-wp/blob/3d1320f43907bd2ff01073cc89cc17929ce20172/includes/Modules/Analytics_4.php#L1003-L1010
```suggestion
												'showProgress' => 'true',
```

hussain-t commented on tests/phpunit/integration/Core/Admin/ScreenTest.php:
```diff
@@ -124,4 +125,199 @@ public function test_enqueue_assets() {
+									'dashboard',
+									array(
+										'slug'         => 'analytics-4',
+										'showProgress' => true,
```
```suggestion
										'showProgress' => 'true',
```

hussain-t commented on tests/phpunit/integration/Core/Admin/ScreenTest.php:
```diff
@@ -124,4 +125,199 @@ public function test_enqueue_assets() {
+			$location = $e->get_location();
+		}
+		$this->assertNotEmpty( $location, 'Redirect should occur.' );
+		$this->assertStringContainsString( 'page=googlesitekit-user-input', $location, 'User input page should be redirected to.' );
```
```suggestion
		$this->assertStringContainsString( 'page=googlesitekit-key-metrics-setup', $location, 'Key metrics setup page should be redirected to.' );
```

hussain-t commented on tests/phpunit/integration/Core/Admin/ScreenTest.php:
```diff
@@ -124,4 +125,199 @@ public function test_enqueue_assets() {
+		}
+		$this->assertNotEmpty( $location, 'Redirect should occur.' );
+		$this->assertStringContainsString( 'page=googlesitekit-user-input', $location, 'User input page should be redirected to.' );
+		$this->assertStringContainsString( 'showProgress=1', $location, 'Show progress should be set.' );
```
```suggestion
		$this->assertStringContainsString( 'showProgress=true', $location, 'Show progress should be set to the string "true".' );
```

hussain-t commented on includes/Core/Admin/Screens.php:
```diff
@@ -359,16 +371,62 @@ private function get_screens() {
+									exit;
+								} else {
+									$slug = $context->input()->filter( INPUT_GET, 'slug' );
+									$show_progress = (bool) $context->input()->filter( INPUT_GET, 'showProgress' );
```
Use `FILTER_VALIDATE_BOOLEAN` when reading the values back
```suggestion
									$show_progress = $context->input()->filter( INPUT_GET, 'showProgress', FILTER_VALIDATE_BOOLEAN );
```

hussain-t commented on includes/Core/Admin/Screens.php:
```diff
@@ -359,16 +371,62 @@ private function get_screens() {
+								} else {
+									$slug = $context->input()->filter( INPUT_GET, 'slug' );
+									$show_progress = (bool) $context->input()->filter( INPUT_GET, 'showProgress' );
+									$re_auth = (bool) $context->input()->filter( INPUT_GET, 'reAuth' );
```
Same here. Use `FILTER_VALIDATE_BOOLEAN` when reading the values back.
```suggestion
									$re_auth = $context->input()->filter( INPUT_GET, 'reAuth', FILTER_VALIDATE_BOOLEAN );
```

hussain-t commented on tests/phpunit/integration/Core/Admin/ScreenTest.php:
```diff
@@ -124,4 +125,199 @@ public function test_enqueue_assets() {
+		$this->assertStringContainsString( 'page=googlesitekit-dashboard', $location, 'Dashboard page should be redirected to.' );
+		$this->assertStringContainsString( 'slug=analytics-4', $location, 'Slug should be set to analytics-4.' );
+		$this->assertStringContainsString( 'showProgress=1', $location, 'Show progress should be set.' );
+		$this->assertStringContainsString( 'reAuth=1', $location, 'Re-auth should be set.' );
```
```suggestion
		$this->assertStringContainsString( 'reAuth=true', $location, 'Re-auth should be set to the string "true".' );
```

hussain-t commented on tests/phpunit/integration/Core/Admin/ScreenTest.php:
```diff
@@ -124,4 +125,199 @@ public function test_enqueue_assets() {
+					if ( ! $is_analytics_connected ) {
+						$slug = $ctx->input()->filter( INPUT_GET, 'slug' );
+						$show_progress = (bool) $ctx->input()->filter( INPUT_GET, 'showProgress' );
+						$re_auth = (bool) $ctx->input()->filter( INPUT_GET, 'reAuth' );
```
```suggestion
						$show_progress = $ctx->input()->filter( INPUT_GET, 'showProgress', FILTER_VALIDATE_BOOLEAN );
						$re_auth = $ctx->input()->filter( INPUT_GET, 'reAuth', FILTER_VALIDATE_BOOLEAN );
```

hussain-t commented on tests/phpunit/integration/Core/Admin/ScreenTest.php:
```diff
@@ -124,4 +125,199 @@ public function test_enqueue_assets() {
+								array(
+									'slug'         => 'analytics-4',
+									'showProgress' => true,
+									'reAuth'       => true,
```
```suggestion
									'showProgress' => 'true',
									'reAuth'       => 'true',
```

techanvil commented on assets/js/components/key-metrics-setup/KeyMetricsSetupApp.test.js:
```diff
@@ -186,17 +196,64 @@ describe( 'KeyMetricsSetupApp', () => {
 			'http://example.com/wp-admin/admin.php?page=googlesitekit-key-metrics-setup';
 
+		fetchMock.postOnce( initialSetupSettingsEndpoint, {
+			body: { initialSetupSettings: true },
```
As above:

```suggestion
			body: { settings: { isAnalyticsSetupComplete: true } },
```

techanvil commented on assets/js/components/setup/SetupUsingProxyWithSignIn/index.test.js:
```diff
@@ -299,6 +307,10 @@ describe( 'SetupUsingProxyWithSignIn', () => {
 
 		it( 'should navigate to the proxy setup URL with Analytics re-auth redirect URL and `showProgress` query argument on CTA click if chosen to connect Analytics', async () => {
+			fetchMock.postOnce( initialSetupSettingsEndpoint, {
+				body: { initialSetupSettings: false },
```
As above:

```suggestion
				body: { settings: { isAnalyticsSetupComplete: false } },
```

techanvil commented on assets/js/components/setup/SetupUsingProxyWithSignIn/index.test.js:
```diff
@@ -377,6 +389,80 @@ describe( 'SetupUsingProxyWithSignIn', () => {
 
+		it( 'should call saveInitialSetupSettings with isAnalyticsSetupComplete: false when starting setup with Analytics', async () => {
+			fetchMock.postOnce( initialSetupSettingsEndpoint, {
+				body: { initialSetupSettings: false },
```
As above:

```suggestion
				body: { settings: { isAnalyticsSetupComplete: false } },
```

techanvil commented on assets/js/components/key-metrics-setup/KeyMetricsSetupApp.test.js:
```diff
@@ -186,17 +196,64 @@ describe( 'KeyMetricsSetupApp', () => {
+			body: { isAnalyticsSetupComplete: null },
+			status: 200,
+		} );
+
```
This isn't needed.

```suggestion
```

techanvil reviewed (changes requested):
Thanks @ankitrox, this is working well but needs a few changes, with a note to be more careful to review AI generated code in future.

AI can often generate what looks at a glance like sensible code but on closer inspection turns out to completely miss the mark, so we must always be vigilant! :)

ankitrox commented:
Thanks for reviewing the changes @techanvil

Sorry about the the phpunit testing changes in wrong file with the wrong strategy as I was not aware we have  `RedirectException` class to test the redirects. I have removed those unnecessary changes in the `tests/phpunit/integration/Core/Admin/ScreenTest.php` and added them to `tests/phpunit/integration/Core/Admin/ScreensTest.php`.

techanvil commented on tests/phpunit/integration/Core/Admin/ScreensTest.php:
```diff
@@ -184,4 +200,106 @@ public function test_menu_order( $given_menu_order, $expected_order ) {
+		$this->screens->register();
+		do_action( 'admin_menu' );
+
+		foreach ( $this->force_get_property( $this->screens, 'screens' ) as $hook_suffix => $screen ) {
```
We can avoid the unused `$screen` variable here:

```suggestion
		foreach ( array_keys( $this->force_get_property( $this->screens, 'screens' ) ) as $hook_suffix ) {
```

techanvil reviewed (approved):
Thanks @ankitrox, that's great. LGTM!

:white_check_mark:
